### PR TITLE
Adds support for Vine.co hrefs

### LINKED ISSUE #47: Add support for Vine

### PR DESCRIPTION
**Pull request by kaganjd**
- Fixes #47 
- Adds context menu for Vine.co hrefs
- Adds getVineId function for send-to-tab
- Adds @6a68's code for getting Vine MP4's from the page source
- Adds get-vine-url.js to /lib

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -115,6 +118,32 @@ cm.Item({
...
+  onMessage: function(url) {
+    const id = getVineId(url);
+    updatePanel({domain: 'vine.co', id: id, src: ''});
+    getVineUrl(url, function(err, streamUrl) {
+      if (!err) updatePanel({src: streamUrl});
+    });
+  }
+});
+
+function getVineId(str) {
```
Comment (meandavejustice) [inline index.js]: This looks like a good opportunity to contribute this function to https://github.com/radiovisual/get-video-id

File: lib/get-vine-url.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+// getVineUrl: Given a vine.co URL, returns URL of the video.
+//
+// Takes two arguments: the vine.co URL and a callback.
+// The callback will be passed two arguments:
+// the first argument is null or an error string,
+// the second argument is the vine CDN URL of the mp4 video (or null).
+function getVineUrl(vineURL, cb) {
+  let hiddenFrame = hiddenFrames.add(hiddenFrames.HiddenFrame({
+    onReady: function() {
+
```
Comment (meandavejustice) [inline lib/get-vine-url.js]: nit: remove this empty space

File: lib/get-vine-url.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+module.exports = getVineUrl;
+
+// getVineUrl: Given a vine.co URL, returns URL of the video.
+//
+// Takes two arguments: the vine.co URL and a callback.
+// The callback will be passed two arguments:
+// the first argument is null or an error string,
+// the second argument is the vine CDN URL of the mp4 video (or null).
+function getVineUrl(vineURL, cb) {
+  let hiddenFrame = hiddenFrames.add(hiddenFrames.HiddenFrame({
```
Comment (meandavejustice) [inline lib/get-vine-url.js]: nit: cc/ @6a68 does this need to be `let` over `const`?

File: lib/get-vine-url.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+var hiddenFrames = require('sdk/frame/hidden-frame');
```
Comment (meandavejustice) [inline lib/get-vine-url.js]: `const` over `var`

File: index.js [hunk 5]
```diff
@@ -115,6 +118,32 @@ cm.Item({
...
+  var id;
+  if (/https:\/\/vine\.co\/v\/([a-zA-Z0-9]*)\/?/.test(str)) {
+    var arr = str.split('/');
```
Comment (meandavejustice) [inline index.js]: use `let` here

**Comment by meandavejustice** [comment]
couple nits, will pull this down tomorrow afternoon(i'm PTO tomorrow morning) for a manual test. This looks good though, thanks @kaganjd

File: lib/get-vine-url.js (hunk 4 above)
Comment (pdehaan) [inline lib/get-vine-url.js]: Humbly submit #102 for your ESLint consideration.

File: lib/get-vine-url.js (hunk 3 above)
Comment (6a68) [inline lib/get-vine-url.js]: const is fine

File: index.js (hunk 1 above)
Comment (kaganjd) [inline index.js]: Done! https://github.com/kaganjd/get-video-id/commit/98f598ee15d63d23d6bb04aab98afe97173067f2

File: lib/get-vine-url.js (hunk 2 above)
Comment (kaganjd) [inline lib/get-vine-url.js]: done

File: lib/get-vine-url.js (hunk 3 above)
Comment (kaganjd) [inline lib/get-vine-url.js]: done

File: lib/get-vine-url.js (hunk 4 above)
Comment (kaganjd) [inline lib/get-vine-url.js]: done

File: index.js (hunk 1 above)
Comment (meandavejustice) [inline index.js]: awesome! let's go ahead an update this pr to bump to version 1.0 of `get-video-id` and remove this function here.

Comment (kaganjd) [inline index.js]: When I download `get-video-id` from npm, it's still version 0.0.6. I'll check again later and update the PR as soon as the version changes.

Comment (pdehaan) [inline index.js]: I just did a `$ npm i get-video-id@latest` in a blank directory and got a `get-video-id@1.0.0` with all the Vine code in the README.

Comment (kaganjd) [inline index.js]: When I do `$ npm i get-video-id@latest`, I get the correct version with code in the correct places. Mysteries... ?

For this, I removed code from index.js, updated the packages.json, packaged a new .xpi, and let the maintainer of the module know about the version thing.

Comment (meandavejustice) [inline index.js]: There isn't an issue with the get-video-id package itself, that is the intended behavior from npm.

If a version is declared in package.json, npm will listen to your configuration and install that version. If you want to install a different version you must specify that either in the package.json or in your install command, hence `npm install get-video-id@1.0.0`

File: index.js [hunk 6]
```diff
@@ -115,6 +118,21 @@ cm.Item({
...
+  label: 'Send to mini player',
+  context: cm.SelectorContext('[href*="vine.co/v/"]'),
+  contentScript: 'self.on("click", function (node, data) {' +
+                 '  self.postMessage(node.href);' +
+                 '});',
+  onMessage: function(url) {
+    const id = getVideoId(url);
+    updatePanel({domain: 'vine.co', id: id, src: ''});
+    getVineUrl(url, function(err, streamUrl) {
+      if (!err) updatePanel({src: streamUrl});
```
Comment (pdehaan) [inline index.js]: Do we need to handle the case where we may have an `err` here?

Comment (6a68) [inline index.js]: Good catch! Actually `getVineUrl` never returns an error message, so we never need to handle it ;-)

Comment (pdehaan) [inline index.js]: Great, thanks. Just wanted to make sure an unexpected error wouldn't cause us to launch a blank video player.

**Comment by kaganjd** [comment]
@meandavejustice there was a merge conflict with the addon so @6a68 suggested we host it somewhere else and delete it out of the repo. I did that and updated the README. If that sounds good to you, you can merge- otherwise, I can make changes.

**Comment by meandavejustice** [comment]
This just needs to have 6b73b37b416df183226b4876b66fd4d65625ca67 removed and get a rebase with master and I'm good to merge :)

**Comment by kaganjd** [comment]
@meandavejustice I think this is done? I'm still a little wobbly on rebases, so I'm not sure.

**Comment by meandavejustice** [comment]
Thanks @kaganjd :tada:

**Pull request merged**